rest-service/manager_rest/rest/resources_v3_1/plugins.py

```diff
@@ -161,18 +162,80 @@ def get(self,
 
 
 class PluginsId(resources_v2_1.PluginsId):
+    @authorize('plugin_upload')
+    @rest_decorators.marshal_with(models.Plugin)
+    def post(self, plugin_id, **kwargs):
+        """Force plugin installation on the given managers or agents.
+
+        This method is for internal use only.
+        """
+        sm = get_storage_manager()
+        action_dict = rest_utils.get_json_and_verify_params({
+            'action': {'type': text_type},
+        })
+        plugin = sm.get(models.Plugin, plugin_id)
+        if action_dict.get('action') == 'install':
+            install_dict = rest_utils.get_json_and_verify_params({
+                'managers': {'type': list},
+                'agents': {'type': list},
+            })
+            get_resource_manager().install_plugin(
+                plugin_id,
+                manager_names=install_dict['managers'],
+                agent_names=install_dict['agents'],
+            )
+            return workflow_executor.install_plugin(plugin)
+        else:
+            raise manager_exceptions.ManagerException(
+                400, 'Unknown action: {0}'.format(action_dict.get('action')))
 
     @authorize('plugin_upload')
     @rest_decorators.marshal_with(models.Plugin)
     def put(self, plugin_id, **kwargs):
-        """
-        For internal use - updates the plugin's installation status.'
-        This method is called from the PluginInstaller after the plugin
-        installation has ended.
-        """
+        """Update the plugin, specifically the installation state.
 
+        Only updating the state is supported right now.
+        This method is for internal use only.
+        """
+        request_dict = rest_utils.get_json_and_verify_params({
+            'agent': {'type': text_type, 'optional': True},
+            'manager': {'type': text_type, 'optional': True},
+            'state': {'type': text_type},
+            'error': {'type': text_type, 'optional': True},
+        })
         sm = get_storage_manager()
+        agent = sm.get(
```

Comment:
Wouldn't it be better to gate checking the agent/manager listing on what was passed in in the request dict to avoid an unnecessary DB lookup on every manager-side install?

Reply:
yeah, rewritten this part